### PR TITLE
manually release ITypeInfo references

### LINKED ISSUE #3271: Flaky test: `IDispatch_GetTypeInfo_Invoke_Success`
**Problem description:**

https://dev.azure.com/dnceng/public/_build/results?buildId=641217&view=logs&j=7c8326b9-0a5f-532a-e6de-db8515c72d9a&s=ff05ad62-bb9a-53b6-ce9f-72f329a63e7c

```
    System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.IDispatchTests.IDispatch_GetTypeInfo_Invoke_Success [FAIL]
      System.Runtime.InteropServices.COMException : Error HRESULT E_FAIL has been returned from a call to a COM component.
      Stack Trace:
           at Interop.Ole32.OleCreatePictureIndirect(PICTDESC* pictdesc, Guid& refiid, BOOL fOwn)
        /_/src/System.Windows.Forms.Primitives/src/Interop/Ole32/Interop.OleCreatePictureIndirect.cs(23,0): at Interop.Ole32.OleCreatePictureIndirect(PICTDESC& pictdesc, Guid& refiid, BOOL fOwn)
        /_/src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs(3988,0): at System.Windows.Forms.AxHost.GetIPictureDispFromPicture(Image image)
        /_/src/System.Windows.Forms.Primitives/tests/Interop/Oleaut32/IDispatchTests.cs(94,0): at System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.IDispatchTests.SubAxHost.GetIPictureDispFromPicture(Image image)
        /_/src/System.Windows.Forms.Primitives/tests/Interop/Oleaut32/IDispatchTests.cs(39,0): at System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.IDispatchTests.IDispatch_GetTypeInfo_Invoke_Success()
```

### LINKED ISSUE #3305: Broken IDispatch.GetTypeInfo interop tests
**.NET Core Version:**
Master with PR #3276 applied

**Have you experienced this same bug with .NET Framework?:**
unknown

**Problem description:**
The [tests](https://github.com/dotnet/winforms/blob/e32620ed185e3c15fa67bdb7fe93757a735d6edb/src/System.Windows.Forms.Primitives/tests/Interop/Oleaut32/ITypeInfoTests.cs) were flaky before, sometimes throwing `ComException(E_FAIL)`. The update of xunit.stafact now is properly hosting everything in an WinForms STA thread and they suddenly seem to fail "reliably" with this error here.
```
System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_AddressOfMember_Invoke_Success
System.Runtime.InteropServices.InvalidComObjectException : COM object that has been separated from its underlying RCW cannot be used. The COM object was released while it was still in use on another thread.
   at System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr ppUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   at Interop.Oleaut32.IDispatch.GetTypeInfo(UInt32 iTInfo, UInt32 lcid, ITypeInfo& ppTInfo)
   at System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_AddressOfMember_Invoke_Success() in D:\projects\winforms\src\System.Windows.Forms.Primitives\tests\Interop\Oleaut32\ITypeInfoTests.cs:line 23

System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_CreateInstance_Invoke_Success
System.Runtime.InteropServices.InvalidComObjectException : COM object that has been separated from its underlying RCW cannot be used. The COM object was released while it was still in use on another thread.
   at System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr ppUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   at Interop.Oleaut32.IDispatch.GetTypeInfo(UInt32 iTInfo, UInt32 lcid, ITypeInfo& ppTInfo)
   at System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_CreateInstance_Invoke_Success() in D:\projects\winforms\src\System.Windows.Forms.Primitives\tests\Interop\Oleaut32\ITypeInfoTests.cs:line 39

System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_GetDllEntry_Invoke_Success
System.Runtime.InteropServices.InvalidComObjectException : COM object that has been separated from its underlying RCW cannot be used. The COM object was released while it was still in use on another thread.
   at System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr ppUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   at Interop.Oleaut32.IDispatch.GetTypeInfo(UInt32 iTInfo, UInt32 lcid, ITypeInfo& ppTInfo)
   at System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_GetDllEntry_Invoke_Success() in D:\projects\winforms\src\System.Windows.Forms.Primitives\tests\Interop\Oleaut32\ITypeInfoTests.cs:line 81

System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_GetFuncDesc_Invoke_Success
System.Runtime.InteropServices.InvalidComObjectException : COM object that has been separated from its underlying RCW cannot be used. The COM object was released while it was still in use on another thread.
   at System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr ppUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   at Interop.Oleaut32.IDispatch.GetTypeInfo(UInt32 iTInfo, UInt32 lcid, ITypeInfo& ppTInfo)
   at System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_GetFuncDesc_Invoke_Success() in D:\projects\winforms\src\System.Windows.Forms.Primitives\tests\Interop\Oleaut32\ITypeInfoTests.cs:line 123

System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_GetIDsOfNames_Invoke_Success
System.Runtime.InteropServices.InvalidComObjectException : COM object that has been separated from its underlying RCW cannot be used. The COM object was released while it was still in use on another thread.
   at System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr ppUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   at Interop.Oleaut32.IDispatch.GetTypeInfo(UInt32 iTInfo, UInt32 lcid, ITypeInfo& ppTInfo)
   at System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_GetIDsOfNames_Invoke_Success() in D:\projects\winforms\src\System.Windows.Forms.Primitives\tests\Interop\Oleaut32\ITypeInfoTests.cs:line 159

System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_GetMops_Invoke_Success
System.Runtime.InteropServices.InvalidComObjectException : COM object that has been separated from its underlying RCW cannot be used. The COM object was released while it was still in use on another thread.
   at System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr ppUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   at Interop.Oleaut32.IDispatch.GetTypeInfo(UInt32 iTInfo, UInt32 lcid, ITypeInfo& ppTInfo)
   at System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_GetMops_Invoke_Success() in D:\projects\winforms\src\System.Windows.Forms.Primitives\tests\Interop\Oleaut32\ITypeInfoTests.cs:line 196

System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_GetRefTypeInfo_Invoke_Success
System.Runtime.InteropServices.InvalidComObjectException : COM object that has been separated from its underlying RCW cannot be used. The COM object was released while it was still in use on another thread.
   at System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr ppUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   at Interop.Oleaut32.IDispatch.GetTypeInfo(UInt32 iTInfo, UInt32 lcid, ITypeInfo& ppTInfo)
   at System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_GetRefTypeInfo_Invoke_Success() in D:\projects\winforms\src\System.Windows.Forms.Primitives\tests\Interop\Oleaut32\ITypeInfoTests.cs:line 236

System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_Invoke_Invoke_Success
System.Runtime.InteropServices.InvalidComObjectException : COM object that has been separated from its underlying RCW cannot be used. The COM object was released while it was still in use on another thread.
   at System.StubHelpers.InterfaceMarshaler.ConvertToManaged(IntPtr ppUnk, IntPtr itfMT, IntPtr classMT, Int32 flags)
   at Interop.Oleaut32.IDispatch.GetTypeInfo(UInt32 iTInfo, UInt32 lcid, ITypeInfo& ppTInfo)
   at System.Windows.Forms.Primitives.Tests.Interop.Oleaut32.ITypeInfoTests.ITypeInfo_Invoke_Invoke_Success() in D:\projects\winforms\src\System.Windows.Forms.Primitives\tests\Interop\Oleaut32\ITypeInfoTests.cs:line 368

```

**Expected behavior:**
Tests pass or there is a reasonable explanation for the error

**Minimal repro:**
Run tests with PR #3276 applied

### PR DESCRIPTION
**Pull request by weltkante**
Fixes #3305
Fixes #3271

## Proposed changes
Release ITypeInfo RCW manually instead of letting the GC clean up

I'm doing this seperately from PR #3276 so it can be reviewed and merged before figuring out all the problems of the package updates.

## Customer Impact
Tests should become more stable regardless how they are run

## Regression? 
Yes (after merging PR #3276)

## Risk
no known risk

### Before
Tests start failing after PR #3276

### After
Tests should succeed before and after PR #3276

## Test methodology
Locally running tests both with and without PR #3276

## Test environment(s)
local

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3315)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/tests/Interop/Oleaut32/ITypeInfoTests.cs [hunk 1]
```diff
@@ -394,5 +411,24 @@ private SubAxHost(string clsidString) : base(clsidString)
 
             public new static IPictureDisp GetIPictureDispFromPicture(Image image) => (IPictureDisp)AxHost.GetIPictureDispFromPicture(image);
         }
+
+        private struct ComRefReleaser : IDisposable
```
Comment (hughbe) [inline src/System.Windows.Forms.Primitives/tests/Interop/Oleaut32/ITypeInfoTests.cs:415]: maybe add a comment or reference. I hate that we have to do this :(

**Comment by hughbe** [comment]
Also, fixes https://github.com/dotnet/winforms/issues/3271

**Comment by weltkante** [comment]
> Also, fixes #3271

#3271 also happens in IPictureTests and I didn't touch those yet. Those tests aren't using ITypeInfo so I'm not sure if #3271 is the same issue. I'd leave it open for now.

**Comment by weltkante** [comment]
Is `ImageCollection_AddStrip_InvokeWithHandle_Success` a new/unknown flaky test? Can't seem to find previews mentions.

**Comment by hughbe** [comment]
Don’t think so. I have noticed some flakiness surrounding ImageList tests with handles in the past

**Comment by weltkante** [comment]
> Just to confirm, is this specific to ITypeInfo?

Its specific to COM objects shared across STA threads (ITypeInfo happens to be shared on the native side according to your linked discussion)

> Surely others will run into this somehow? Does this happen in netfx

You need to have mutliple STA threads starting up/shutting down, this is not a common scenario, and if you have it you run into a lot of other problems (for example SystemEvents not dealing gracefully with it, WinForms itself doesn't deal gracefully with it, WPF has a few issues posted). From what I can tell having multiple STA threads starting and terminating needs more work across the whole framework before it can work reliably. And I don't think the old COM interop is going to get any fixes in this area.

> I think we should somehow have tests in dotnet/runtime verifying this behaviour.

The best you could do is verify that the bug still exists? I think the point is that the behavior is so complex and edge case that its not worth to fix. Writing a test to ensure something is not fixed is odd.

**Comment by hughbe** [comment]
Thanks, is the concern around the fix because it would break things or just that we don’t want to fix anything because it’s too much effort?

**Comment by weltkante** [comment]
The problem is probably to fix _only_ the bug without introducing regressions. You need to change the way COM objects are associated to apartments. I had commented an idea on the other thread:

> I guess the right fix for this bug in .NET would be to drop the apartment association in the RCW cache if you see the same COM identity being used from multiple apartments without proxy

There may be existing code which relies on the "wrong" apartment association for some side effect.

Introducing/replacing functionality like this is always risky to get some edge case wrong. I have the impression they'd rather spend their time on creating the new COM interop system which will be more open to taking such fixes.

**Comment by RussKie** [comment]
Hmm, now we have `ImageCollection_AddStrip_InvokeWithHandle_Success` test failing...
```
    System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_AddStrip_InvokeWithHandle_Success(transparentColor: Color [A=18, R=52, G=86, B=120], value: Bitmap { Flags = 2, FrameDimensionsList = [7462dc86-6180-4c7e-8e3f-ee7333a7a483], Height = 16, HorizontalResolution = 96, Palette = ColorPalette { Entries = [...], Flags = 0 }, ... }, expectedCount: 1) [FAIL]
      System.InvalidOperationException : Image cannot be added to the ImageList.
      Stack Trace:
        F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\ImageList.cs(414,0): at System.Windows.Forms.ImageList.AddToHandle(Bitmap bitmap)
        F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\ImageList.ImageCollection.cs(336,0): at System.Windows.Forms.ImageList.ImageCollection.Add(Original original, ImageInfo imageInfo)
        F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\ImageList.ImageCollection.cs(427,0): at System.Windows.Forms.ImageList.ImageCollection.AddStrip(Image value)
        F:\workspace\_work\1\s\src\System.Windows.Forms\tests\UnitTests\System\Windows\Forms\ImageList.ImageCollectionTests.cs(881,0): at System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_AddStrip_InvokeWithHandle_Success(Color transparentColor, Image value, Int32 expectedCount)
```

**Comment by weltkante** [comment]
> Hmm, now we have ImageCollection_AddStrip_InvokeWithHandle_Success test failing...

Yeah I mentioned that above, seems its not know to be flaky, but it can't have anything to do with this PR. This is literally just cleaning up manually instead of letting the GC do it.

**Comment by RussKie** [comment]
Looks like the first timer... Passed on a rebuild.

**Comment by RussKie** [comment]
Thank you for the fix! 🚀
